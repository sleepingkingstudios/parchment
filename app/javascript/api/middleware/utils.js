import { convertToArray } from '../../utils/array';

const mapProperty = (ary, propName) => {
  if (typeof propName === 'undefined' || propName === null) { return ary; }

  return ary.map(item => item[propName]);
};

const injectMiddlewareAfter = (ary, middleware, target) => {
  const copy = ary.slice();

  if (target === ':all') {
    copy.splice(target.length, 0, middleware);

    return copy;
  }

  const index = ary.findIndex(item => (item.type === target));

  if (index === -1) {
    throw new Error(`Unable to inject middleware after "${target}"`);
  }

  copy.splice(index + 1, 0, middleware);

  return copy;
};

const injectMiddlewareBefore = (ary, middleware, target) => {
  const copy = ary.slice();

  if (target === ':all') {
    copy.splice(0, 0, middleware);

    return copy;
  }

  const index = ary.findIndex(item => (item.type === target));

  if (index === -1) {
    throw new Error(`Unable to inject middleware before "${target}"`);
  }

  copy.splice(index, 0, middleware);

  return copy;
};

const applyDirective = (ary, directive) => {
  const { middleware } = directive;

  if (directive.after) {
    return injectMiddlewareAfter(ary, middleware, directive.after);
  }

  if (directive.before) {
    return injectMiddlewareBefore(ary, middleware, directive.before);
  }

  return ary;
};

export const applyMiddleware = (original, middleware) => {
  if (middleware.length === 0) { return original; }

  let fn = original;

  middleware.reverse().forEach((middlewareFn) => { fn = middlewareFn(fn); });

  return fn;
};

export const injectMiddleware = (middleware, directives) => (
  directives.reduce(applyDirective, middleware)
);

export const selectMiddleware = (maybeMiddleware, propName) => {
  const middleware = convertToArray(maybeMiddleware);
  const selected = [];

  mapProperty(middleware, propName).forEach((item) => {
    if (typeof item === 'function') { selected.push(item); }
  });

  return selected;
};

export const wrapMiddleware = (middleware, next) => {
  if (typeof middleware !== 'function') { return next; }

  return middleware(next);
};
