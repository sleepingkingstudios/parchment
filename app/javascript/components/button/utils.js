export const buttonStyleClass = ({ buttonStyle, outline }) => {
  switch (buttonStyle) {
    case 'link':
      return 'btn-link';
    case 'primary':
    case 'secondary':
    case 'success':
    case 'danger':
    case 'warning':
    case 'info':
    case 'light':
    case 'dark':
      return `btn-${outline ? 'outline-' : ''}${buttonStyle}`;
    default:
      return `btn-${outline ? 'outline-' : ''}primary`;
  }
};

export const buttonClass = ({
  block,
  buttonStyle,
  className,
  outline,
}) => {
  const classes = ['btn', buttonStyleClass({ buttonStyle, outline })];

  if (block) { classes.push('btn-block'); }

  if (className) { classes.push(className); }

  return classes.join(' ');
};

export const disabledOnClick = ({ preventDefault }) => { preventDefault(); };
