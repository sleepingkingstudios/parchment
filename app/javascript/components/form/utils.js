import {
  convertToArray,
} from '../../utils/array';
import {
  camelize,
  underscore,
} from '../../utils/string';

export const formatErrors = (errors) => {
  if (typeof errors === 'undefined') { return {}; }
  if (errors === null) { return {}; }

  const formatted = {};

  errors.forEach((tuple) => {
    const property = camelize(tuple[0]);
    const message = tuple[1];

    if (!formatted[property]) { formatted[property] = []; }

    formatted[property].push(message);
  });

  return formatted;
};

export const generateFieldId = ({ inputId, path, prop }) => {
  if (inputId) { return inputId; }

  const segments = convertToArray(path).slice();

  segments.push(prop);

  segments.push('input');

  return segments
    .map(str => underscore(str).replace(/_/g, '-'))
    .join('-');
};

export const getInputValue = ({ checked, type, value }) => {
  if (type === 'checkbox') { return checked; }

  return value;
};

export const coerceInputValue = (handler, fn) => (
  (event) => {
    const { target } = event;
    const value = fn(getInputValue(target));

    return handler({ target: { value } });
  }
);
