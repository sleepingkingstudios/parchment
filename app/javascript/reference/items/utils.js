import {
  capitalize,
  underscore,
} from 'utils/string';

// eslint-disable-next-line import/prefer-default-export
export const formatType = (str) => {
  if (typeof str === 'undefined' || str == null) { return ''; }

  const segments = str.split('::');

  return underscore(segments[segments.length - 1])
    .split(/[_ ]/)
    .map(capitalize)
    .join(' ');
};
