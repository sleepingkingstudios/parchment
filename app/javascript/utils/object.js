import {
  camelize,
  underscore,
} from './string';

const copyObjectOrArray = (obj) => {
  if (Array.isArray(obj)) { return obj.slice(); }

  return Object.assign({}, obj);
};

const guardArrayIndex = (ary, index) => {
  if (!Array.isArray(ary) || Number.isInteger(index)) { return; }

  throw new Error(`invalid Array index ${index}`);
};

const isPrimitive = (obj) => {
  if (obj === null || obj === true || obj === false) { return true; }

  const type = typeof obj;

  if (type === 'string' || type === 'number') { return true; }

  return false;
};

const maybeInteger = (key, defaultValue) => {
  if (typeof key === 'undefined' || key === null) { return defaultValue; }

  const int = parseInt(key, 10);

  return (int.toString() === key) ? int : key;
};

export const exists = obj => !(obj === null || typeof obj === 'undefined');

const recursivelyMapKeys = (obj, fn) => {
  if (!exists(obj)) { return obj; }

  if (isPrimitive(obj)) { return obj; }

  if (Array.isArray(obj)) {
    return obj.map(item => recursivelyMapKeys(item, fn));
  }

  const mapped = {};

  Object.entries(obj).forEach(([key, value]) => {
    mapped[fn(key)] = recursivelyMapKeys(value, fn);
  });

  return mapped;
};

export const assign = (obj, value, ...path) => {
  if (path.length === 0) { throw new Error("path can't be blank"); }

  const last = path.pop();
  const root = copyObjectOrArray(obj);
  let current = root;

  path.forEach((segment, index) => {
    const key = maybeInteger(segment);
    const next = maybeInteger(path[index + 1], last);

    if (Object.prototype.hasOwnProperty.call(current, key)) {
      guardArrayIndex(current, key);

      current[key] = copyObjectOrArray(current[key]);
    } else {
      current[key] = Number.isInteger(next) ? [] : {};
    }

    current = current[key];
  });

  const key = maybeInteger(last);

  guardArrayIndex(current, key);

  current[key] = value;

  return root;
};

export const camelizeKeys = obj => (
  exists(obj) ? recursivelyMapKeys(obj, camelize) : {}
);

export const underscoreKeys = obj => (
  exists(obj) ? recursivelyMapKeys(obj, underscore) : {}
);

export const valueOrDefault = (obj, defaultValue = null) => (
  exists(obj) ? obj : defaultValue
);

export const dig = (obj, ...path) => {
  if (path.length === 0) { return obj; }

  const childName = path[0];

  if (!Object.prototype.hasOwnProperty.call(obj, childName)) { return null; }

  return dig(obj[childName], ...path.slice(1));
};
