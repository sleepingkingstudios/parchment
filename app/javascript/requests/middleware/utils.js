import { convertToArray } from '../../utils/array';

const mapProperty = (ary, propName) => {
  if (typeof propName === 'undefined' || propName === null) { return ary; }

  return ary.map(item => item[propName]);
};

export const applyMiddleware = (original, middleware) => {
  if (middleware.length === 0) { return original; }

  let fn = original;

  middleware.reverse().forEach((middlewareFn) => { fn = middlewareFn(fn); });

  return fn;
};

export const selectMiddleware = (maybeMiddleware, propName) => {
  const middleware = convertToArray(maybeMiddleware);
  const selected = [];

  mapProperty(middleware, propName).forEach((item) => {
    if (typeof item === 'function') { selected.push(item); }
  });

  return selected;
};
