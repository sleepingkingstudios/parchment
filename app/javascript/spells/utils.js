import { capitalize } from '../utils/string';
import { shortOrdinal } from '../utils/number';

export const formatComponents = (spell, includeMaterials = false) => {
  const components = [];

  if (spell.verbalComponent) { components.push('V'); }

  if (spell.somaticComponent) { components.push('S'); }

  if ((spell.materialComponent || 0).length > 0) {
    if (includeMaterials) {
      components.push(`M (${spell.materialComponent})`);
    } else {
      components.push('M');
    }
  }

  if (components.length === 0) { return 'None'; }

  return components.join(', ');
};

export const formatSchoolAndLevel = ({ level, ritual, school }) => {
  const fragments = [];

  if (level === 0) {
    fragments.push(capitalize(school));
    fragments.push('cantrip');
  } else {
    fragments.push(`${shortOrdinal(level)}-level`);
    fragments.push(school.toLowerCase());
  }

  if (ritual) { fragments.push('(ritual)'); }

  return fragments.join(' ');
};
