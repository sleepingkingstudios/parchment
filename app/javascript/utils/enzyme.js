import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
