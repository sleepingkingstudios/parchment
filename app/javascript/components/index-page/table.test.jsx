import React from 'react';
import { shallow } from 'enzyme';

import IndexPageTable from './table';
import {
  INITIALIZED,
  PENDING,
  FAILURE,
  SUCCESS,
} from '../../api/status';
import { valueOrDefault } from '../../utils/object';

const buildEndpoint = ({ performRequest, state }) => ({
  hooks: {
    useEndpoint: () => state,
  },
  request: {
    performRequest: valueOrDefault(performRequest, () => {}),
  },
});

describe('<IndexPageTable />', () => {
  const columns = [
    {
      label: 'Name',
      prop: 'name',
    },
  ];
  const resourceName = 'Widget';
  const defaultProps = { columns, resourceName };

  describe('with status: INITIALIZED', () => {
    const state = { data: {}, status: INITIALIZED };

    it('should display the pending message', () => {
      const endpoint = buildEndpoint({ state });
      const wrapper = shallow(<IndexPageTable {...defaultProps} endpoint={endpoint} />);
      const rendered = wrapper.find('StatusSwitch').renderProp('renderInitialized')();
      const { status } = state;

      expect(wrapper).toHaveDisplayName('StatusSwitch');
      expect(wrapper).toHaveProp({ status });

      expect(rendered).toHaveDisplayName('p');
      expect(rendered).toHaveText('Loading widgets data from the server...');
    });
  });

  describe('with status: FAILURE', () => {
    const state = { data: {}, status: FAILURE };

    it('should display the failure message', () => {
      const endpoint = buildEndpoint({ state });
      const wrapper = shallow(<IndexPageTable {...defaultProps} endpoint={endpoint} />);
      const rendered = wrapper.find('StatusSwitch').renderProp('renderFailure')();
      const { status } = state;

      expect(wrapper).toHaveDisplayName('StatusSwitch');
      expect(wrapper).toHaveProp({ status });

      expect(rendered).toHaveDisplayName('p');
      expect(rendered).toHaveText('Unable to load widgets data from the server.');
    });
  });

  describe('with status: PENDING', () => {
    const state = { data: {}, status: PENDING };

    it('should display the pending message', () => {
      const endpoint = buildEndpoint({ state });
      const wrapper = shallow(<IndexPageTable {...defaultProps} endpoint={endpoint} />);
      const rendered = wrapper.find('StatusSwitch').renderProp('renderInitialized')();
      const { status } = state;

      expect(wrapper).toHaveDisplayName('StatusSwitch');
      expect(wrapper).toHaveProp({ status });

      expect(rendered).toHaveDisplayName('p');
      expect(rendered).toHaveText('Loading widgets data from the server...');
    });
  });

  describe('with status: SUCCESS', () => {
    const widgets = [
      { name: 'Widget' },
      { name: 'Gadget' },
      { name: 'Thingamabob' },
    ];
    const state = { data: { widgets }, status: SUCCESS };

    it('should render the table', () => {
      const endpoint = buildEndpoint({ state });
      const { status } = state;
      const wrapper = shallow(<IndexPageTable {...defaultProps} endpoint={endpoint} />);
      const rendered = wrapper
        .find('StatusSwitch')
        .renderProp('renderSuccess')({ widgets });
      const message = 'There are no widgets matching the criteria.';

      expect(wrapper).toHaveDisplayName('StatusSwitch');
      expect(wrapper).toHaveProp({ status });

      expect(rendered).toHaveDisplayName('Table');
      expect(rendered).toHaveProp({ columns });
      expect(rendered).toHaveProp({ data: widgets });
      expect(rendered).toHaveProp({ message });
    });

    it('should pass the onDelete handler to the table', () => {
      const inner = jest.fn();
      const performRequest = () => inner;
      const endpoint = buildEndpoint({ performRequest, state });
      const wrapper = shallow(<IndexPageTable {...defaultProps} endpoint={endpoint} />);
      const rendered = wrapper
        .find('StatusSwitch')
        .renderProp('renderSuccess')({ widgets });
      const handler = rendered.prop('cellProps').onDelete;
      const dispatch = jest.fn();
      const getState = jest.fn();

      expect(typeof handler).toEqual('function');

      handler({ dispatch, getState });

      expect(inner).toHaveBeenCalledWith(dispatch, getState);
    });

    it('should match the snapshot', () => {
      const endpoint = buildEndpoint({ state });
      const rendered = shallow(<IndexPageTable {...defaultProps} endpoint={endpoint} />);

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with mapData: function', () => {
    const widgets = [
      { name: 'Widget' },
      { name: 'Gadget' },
      { name: 'Thingamabob' },
    ];
    const state = { data: { archived: { widgets } }, status: SUCCESS };
    const mapData = data => data.archived.widgets;

    it('should render the Spells table', () => {
      const endpoint = buildEndpoint({ state });
      const { status } = state;
      const wrapper = shallow(
        <IndexPageTable {...defaultProps} endpoint={endpoint} mapData={mapData} />,
      );
      const rendered = wrapper
        .find('StatusSwitch')
        .renderProp('renderSuccess')({ widgets });
      const message = 'There are no widgets matching the criteria.';

      expect(wrapper).toHaveDisplayName('StatusSwitch');
      expect(wrapper).toHaveProp({ status });

      expect(rendered).toHaveDisplayName('Table');
      expect(rendered).toHaveProp({ columns });
      expect(rendered).toHaveProp({ data: widgets });
      expect(rendered).toHaveProp({ message });
    });
  });

  describe('with Table: component', () => {
    const widgets = [
      { name: 'Widget' },
      { name: 'Gadget' },
      { name: 'Thingamabob' },
    ];
    const state = { data: { widgets }, status: SUCCESS };
    const CustomTable = () => (<div />);

    it('should render the table', () => {
      const endpoint = buildEndpoint({ state });
      const { status } = state;
      const wrapper = shallow(
        <IndexPageTable {...defaultProps} endpoint={endpoint} Table={CustomTable} />,
      );
      const rendered = wrapper
        .find('StatusSwitch')
        .renderProp('renderSuccess')({ widgets });
      const message = 'There are no widgets matching the criteria.';

      expect(wrapper).toHaveDisplayName('StatusSwitch');
      expect(wrapper).toHaveProp({ status });

      expect(rendered).toHaveDisplayName('CustomTable');
      expect(rendered).toHaveProp({ columns });
      expect(rendered).toHaveProp({ data: widgets });
      expect(rendered).toHaveProp({ message });
    });
  });
});
