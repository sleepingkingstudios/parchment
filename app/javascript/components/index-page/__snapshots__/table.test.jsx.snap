// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexPageTable /> with status: SUCCESS should match the snapshot 1`] = `
<StatusSwitch
  renderFailure={[Function]}
  renderInitialized={[Function]}
  renderPending={[Function]}
  renderSuccess={[Function]}
  status="success"
/>
`;
